### PR TITLE
[Feature-3134]Refactor to use a shared singleton Curator Zookeeper Client

### PR DESCRIPTION
**Pull request by tswstarplanet**
#3134 
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds checkstyle plugin.)*

## Brief change log

*(for example:)*
  - Add a CuratorZookeeperClient class to hold a shared CuratorFramework instance.
## Verify this pull request

*(Please pick either of the following options)*

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added dolphinscheduler-dao tests for end-to-end.*
  - *Added CronUtilsTest to verify the change.*
  - *Manually verified the change by testing locally.*

### CODE REVIEW

**Comment by yangyichao-mango** [comment]
Good job,
Is there a better way to to complete the singleton mode to avoid multithreading problems and improve the robust of code?

[1] https://www.geeksforgeeks.org/singleton-design-pattern/

**Comment by tswstarplanet** [comment]
I use the concept of singleton here just to emphasize avoiding create multiple Curator Zookeeper Client. Here it's suitable to use the @Component to implement the singleton in Spring environment.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/CuratorZookeeperClient.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        }
+        zkClient = builder.build();
+        zkClient.start();
+        try {
+            zkClient.blockUntilConnected();
+        } catch (final Exception ex) {
+            throw new RuntimeException(ex);
+        }
+        return zkClient;
+    }
```
Comment (yangyichao-mango) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/CuratorZookeeperClient.java:92]:
```suggestion
    private volatile static CuratorFramework zkClient;

    @Override
    public void afterPropertiesSet() throws Exception {
        zkClient = buildClient();
        initStateLister();
    }

    private CuratorFramework buildClient() {

        if (null == zkClient) {
            synchronized (CuratorZookeeperClient.class) {
                if (null == zkClient) {
                    logger.info("zookeeper registry center init, server lists is: {}.", zookeeperConfig.getServerList());

                    CuratorFrameworkFactory.Builder builder = CuratorFrameworkFactory.builder().ensembleProvider(new DefaultEnsembleProvider(checkNotNull(zookeeperConfig.getServerList(),"zookeeper quorum can't be null")))
                            .retryPolicy(new ExponentialBackoffRetry(zookeeperConfig.getBaseSleepTimeMs(), zookeeperConfig.getMaxRetries(), zookeeperConfig.getMaxSleepMs()));

                    //these has default value
                    if (0 != zookeeperConfig.getSessionTimeoutMs()) {
                        builder.sessionTimeoutMs(zookeeperConfig.getSessionTimeoutMs());
                    }
                    if (0 != zookeeperConfig.getConnectionTimeoutMs()) {
                        builder.connectionTimeoutMs(zookeeperConfig.getConnectionTimeoutMs());
                    }
                    if (StringUtils.isNotBlank(zookeeperConfig.getDigest())) {
                        builder.authorization("digest", zookeeperConfig.getDigest().getBytes(StandardCharsets.UTF_8)).aclProvider(new ACLProvider() {

                            @Override
                            public List<ACL> getDefaultAcl() {
                                return ZooDefs.Ids.CREATOR_ALL_ACL;
                            }

                            @Override
                            public List<ACL> getAclForPath(final String path) {
                                return ZooDefs.Ids.CREATOR_ALL_ACL;
                            }
                        });
                    }
                    zkClient = builder.build();
                    zkClient.start();
                    try {
                        zkClient.blockUntilConnected();
                    } catch (final Exception ex) {
                        throw new RuntimeException(ex);
                    }
                }
            }
        }
        return zkClient;
    }
```

**Comment by dailidong** [comment]
> SonarCloud Quality Gate failed.
> 

SonarCloud Quality Gate failed.

**Comment by tswstarplanet** [comment]
> > SonarCloud Quality Gate failed.
> 
> SonarCloud Quality Gate failed.

To make unit test effective, it's need to use the ZKServer that are refactor by pr #3141

**Comment by qiaozhanwei** [comment]
please resolve conflicting files and ut coverage rate. Thx

**Pull request closed without merging**